File: src/dict.js
/**
 * AICL Dictionary loader
 *
 * Loads the three dictionary files + modifiers, merges them into a single lookup,
 * and builds both the forward (pattern -> symbol) and reverse
 * (symbol -> pattern) maps. Patterns are sorted by length for
 * longest-match-first encoding.
 */

import { readFileSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import { codePoints } from './unicode.js';

const __dirname = dirname(fileURLToPath(import.meta.url));
const DICT_DIR = join(__dirname, '..', 'dict');

/** Cache so we only load/build once per process. */
let cache = null;

/**
 * Load the raw dictionary objects from dict/*.json
 * @returns {{ english: object, code: object, symbols: object, modifiers: object }}
 */
export function loadRaw() {
  return {
    english: JSON.parse(readFileSync(join(DICT_DIR, 'english.json'), 'utf-8')),
    code: JSON.parse(readFileSync(join(DICT_DIR, 'code.json'), 'utf-8')),
    symbols: JSON.parse(readFileSync(join(DICT_DIR, 'symbols.json'), 'utf-8')),
    modifiers: JSON.parse(readFileSync(join(DICT_DIR, 'modifiers.json'), 'utf-8')),
  };
}

/**
 * Modifier transform functions — maps modifier name to (word) => transformed word.
 * These reconstruct the original text from base words during decoding.
 */
const MODIFIER_TRANSFORMS = {
  MOD_CAPS:          (w) => w.charAt(0).toUpperCase() + w.slice(1),
  MOD_TRAIL_SPACE:   (w) => w + ' ',
  MOD_TRAIL_COMMA:   (w) => w + ',',
  MOD_TRAIL_PERIOD:  (w) => w + '.',
  MOD_TRAIL_QUESTION:(w) => w + '?',
  MOD_TRAIL_EXCL:    (w) => w + '!',
  MOD_TRAIL_SEMI:    (w) => w + ';',
  MOD_TRAIL_COLON:   (w) => w + ':',
  MOD_TRAIL_RPAREN:  (w) => w + ')',
  MOD_TRAIL_RBRACKET:(w) => w + ']',
  MOD_TRAIL_RBRACE:  (w) => w + '}',
  MOD_TRAIL_RQUOTE:  (w) => w + '"',
  MOD_LEAD_SPACE:    (w) => ' ' + w,
  MOD_LEAD_LPAREN:   (w) => '(' + w,
  MOD_LEAD_LBRACKET: (w) => '[' + w,
  MOD_LEAD_LBRACE:   (w) => '{' + w,
  MOD_LEAD_LQUOTE:   (w) => '"' + w,
};

/**
 * Build the merged, sorted lookup structures.
 * @returns {{
 *   patternToSymbol: Map<string,string>,
 *   symbolToPattern: Map<string,string>,
 *   sortedPatterns: string[],
 *   modifierSymbols: Set<string>,   // set of modifier code point chars
 *   modifierMap: Map<string,{name:string, transform:function}>,  // symbol -> {name, transform}
 *   size: number
 * }}
 */
export function build() {
  if (cache) return cache;

  const { english, code, symbols, modifiers } = loadRaw();
  const patternToSymbol = new Map();
  const symbolToPattern = new Map();

  for (const dict of [english, code, symbols]) {
    for (const [pattern, symbol] of Object.entries(dict)) {
      if (!patternToSymbol.has(pattern)) {
        patternToSymbol.set(pattern, symbol);
        symbolToPattern.set(symbol, pattern);
      }
    }
  }

  // Build modifier lookup: symbol char -> { name, transform }
  const modifierSymbols = new Set();
  const modifierMap = new Map();
  for (const [name, symbol] of Object.entries(modifiers)) {
    modifierSymbols.add(symbol);
    modifierMap.set(symbol, {
      name,
      transform: MODIFIER_TRANSFORMS[name] || ((w) => w),
    });
    // Also add to patternToSymbol so encoder can find modifiers by name
    if (!patternToSymbol.has(name)) {
      patternToSymbol.set(name, symbol);
      symbolToPattern.set(symbol, name);
    }
  }

  // Longest first so greedy matching finds the best (longest) hit.
  const sortedPatterns = [...patternToSymbol.keys()].sort(
    (a, b) => b.length - a.length || (a < b ? -1 : 1)
  );

  cache = {
    patternToSymbol,
    symbolToPattern,
    sortedPatterns,
    modifierSymbols,
    modifierMap,
    size: patternToSymbol.size,
    english,
    code,
    symbols,
    modifiers,
  };

  return cache;
}

/** Check if a code point char is a modifier symbol. */
export function isModifier(ch) {
  const { modifierSymbols } = build();
  return modifierSymbols.has(ch);
}

/** Get the modifier transform for a symbol char. Returns null if not a modifier. */
export function getModifierTransform(ch) {
  const { modifierMap } = build();
  const mod = modifierMap.get(ch);
  return mod ? mod.transform : null;
}

/** Invalidate the cache (mainly for tests). */
export function reset() {
  cache = null;
}

/**
 * Get a single dict's entries (convenience for tooling/inspection).
 * @returns {{ english: object, code: object, symbols: object, modifiers: object }}
 */
export function dicts() {
  return loadRaw();
}

File: aicl/dict.py
"""Dictionary loading and immutable runtime indexes for AICL stage 1."""

from __future__ import annotations

import json
from functools import lru_cache
from importlib import resources
from pathlib import Path
from types import MappingProxyType
from typing import Any, Callable, Final, Mapping

from .unicode import _utf16_length

_MODIFIER_TRANSFORMS: Final[dict[str, Callable[[str], str]]] = {
    "MOD_CAPS": lambda word: word[:1].upper() + word[1:],
    "MOD_ALLCAPS": str.upper,
    "MOD_TRAIL_SPACE": lambda word: word + " ",
    "MOD_TRAIL_COMMA": lambda word: word + ",",
    "MOD_TRAIL_PERIOD": lambda word: word + ".",
    "MOD_TRAIL_QUESTION": lambda word: word + "?",
    "MOD_TRAIL_EXCL": lambda word: word + "!",
    "MOD_TRAIL_SEMI": lambda word: word + ";",
    "MOD_TRAIL_COLON": lambda word: word + ":",
    "MOD_TRAIL_RPAREN": lambda word: word + ")",
    "MOD_TRAIL_RBRACKET": lambda word: word + "]",
    "MOD_TRAIL_RBRACE": lambda word: word + "}",
    "MOD_TRAIL_RQUOTE": lambda word: word + '"',
    "MOD_LEAD_SPACE": lambda word: " " + word,
    "MOD_LEAD_LPAREN": lambda word: "(" + word,
    "MOD_LEAD_LBRACKET": lambda word: "[" + word,
    "MOD_LEAD_LBRACE": lambda word: "{" + word,
    "MOD_LEAD_LQUOTE": lambda word: '"' + word,
}


class TrieNode:
    """A compact trie node used by the encoder's longest-match search."""

    __slots__ = ("children", "symbol")

    def __init__(self) -> None:
        self.children: dict[str, TrieNode] = {}
        self.symbol: str | None = None


_REPOSITORY_ROOT = Path(__file__).resolve().parents[1]


def _resource_json(filename: str) -> Any:
    source = _REPOSITORY_ROOT / "dict" / filename
    if source.is_file():
        return json.loads(source.read_text(encoding="utf-8"))
    resource = resources.files("aicl.data").joinpath(filename)
    return json.loads(resource.read_text(encoding="utf-8"))


@lru_cache(maxsize=1)
def _cached_raw() -> tuple[dict[str, str], dict[str, str], dict[str, str], dict[str, str]]:
    return (
        _resource_json("english.json"),
        _resource_json("code.json"),
        _resource_json("symbols.json"),
        _resource_json("modifiers.json"),
    )


def load_raw() -> dict[str, dict[str, str]]:
    """Load a defensive copy of the four dictionary JSON objects."""

    english, code, symbols, modifiers = _cached_raw()
    return {
        "english": dict(english),
        "code": dict(code),
        "symbols": dict(symbols),
        "modifiers": dict(modifiers),
    }


def _build_trie(pattern_to_symbol: Mapping[str, str]) -> TrieNode:
    root = TrieNode()
    for pattern, symbol in pattern_to_symbol.items():
        if pattern.startswith("MOD_"):
            continue
        node = root
        for ch in pattern:
            child = node.children.get(ch)
            if child is None:
                child = TrieNode()
                node.children[ch] = child
            node = child
        node.symbol = symbol
    return root


def _sorted_patterns(pattern_to_symbol: Mapping[str, str]) -> list[str]:
    # JavaScript's Array.sort comparator uses UTF-16 string length.  The
    # dictionary is ASCII today, but using the compatibility helper keeps the
    # ordering rule exact if a supplementary pattern is added later.
    return sorted(pattern_to_symbol, key=lambda pattern: (-_utf16_length(pattern), pattern))


_CACHE: dict[str, object] | None = None


def build() -> Mapping[str, object]:
    """Return the process-wide, cached dictionary indexes.

    The returned mapping has the same logical fields as the JavaScript build
    object.  Its maps and raw dictionaries are read-only views; callers that
    need to modify data can use :func:`dicts` to obtain fresh copies.
    """

    global _CACHE
    if _CACHE is not None:
        return MappingProxyType(_CACHE)

    english, code, symbols, modifiers = _cached_raw()
    pattern_to_symbol: dict[str, str] = {}
    symbol_to_pattern: dict[str, str] = {}
    for dictionary in (english, code, symbols):
        for pattern, symbol in dictionary.items():
            if pattern not in pattern_to_symbol:
                pattern_to_symbol[pattern] = symbol
                symbol_to_pattern[symbol] = pattern

    modifier_symbols: set[str] = set()
    modifier_map: dict[str, dict[str, object]] = {}
    for name, symbol in modifiers.items():
        modifier_symbols.add(symbol)
        modifier_map[symbol] = MappingProxyType(
            {
                "name": name,
                "transform": _MODIFIER_TRANSFORMS.get(name, lambda word: word),
            }
        )
        if name not in pattern_to_symbol:
            pattern_to_symbol[name] = symbol
            symbol_to_pattern[symbol] = name

    sorted_patterns = _sorted_patterns(pattern_to_symbol)
    fragment_patterns = [
        pattern
        for pattern in sorted_patterns
        if 2 <= _utf16_length(pattern) <= 4
    ]

    _CACHE = {
        "patternToSymbol": MappingProxyType(pattern_to_symbol),
        "symbolToPattern": MappingProxyType(symbol_to_pattern),
        "sortedPatterns": tuple(sorted_patterns),
        "modifierSymbols": frozenset(modifier_symbols),
        "modifierMap": MappingProxyType(modifier_map),
        "size": len(pattern_to_symbol),
        "english": MappingProxyType(english),
        "code": MappingProxyType(code),
        "symbols": MappingProxyType(symbols),
        "modifiers": MappingProxyType(modifiers),
        "trie": _build_trie(pattern_to_symbol),
        "fragmentPatterns": tuple(fragment_patterns),
    }
    return MappingProxyType(_CACHE)


def is_modifier(ch: str) -> bool:
    """Return whether *ch* is a modifier symbol."""

    return ch in build()["modifierSymbols"]  # type: ignore[operator]


def get_modifier_transform(ch: str) -> Callable[[str], str] | None:
    """Return the transform associated with a modifier, or ``None``."""

    modifier = build()["modifierMap"].get(ch)  # type: ignore[union-attr]
    if modifier is None:
        return None
    return modifier["transform"]  # type: ignore[return-value]


def reset() -> None:
    """Invalidate the process-wide dictionary index (primarily for tests)."""

    global _CACHE
    _CACHE = None
    _cached_raw.cache_clear()


def dicts() -> dict[str, dict[str, str]]:
    """Return defensive copies of the individual raw dictionaries."""

    return load_raw()


# Public JavaScript-shaped and Python-shaped aliases.
build_dict = build
reset_dict = reset
buildDict = build
resetDict = reset
isModifier = is_modifier
getModifierTransform = get_modifier_transform

__all__ = [
    "TrieNode",
    "build",
    "buildDict",
    "build_dict",
    "dicts",
    "getModifierTransform",
    "get_modifier_transform",
    "isModifier",
    "is_modifier",
    "load_raw",
    "reset",
    "resetDict",
    "reset_dict",
]
